### ISSUE TITLE
How does Adagrad work in keras?

### ISSUE THREAD

**Issue by zhengsx** [fchollet/keras#5125]
Hi all,

I post a question about keras' optimizer in [stackoverflow](http://stackoverflow.com/questions/41787873/how-adagrad-wroks-in-keras-what-does-self-weights-mean-in-keras-optimizer).  Can anybody help me to answer that?

The function `get_update()` in [Adagrad](https://github.com/fchollet/keras/blob/master/keras/optimizers.py#L288) seems one step update. However should the accumulators be stored the history information? Why it has been initialized to zeros at each step? How it can be an accumulator through the whole training process?

What does this line do?
`self.weights = accumulators`
It seems self.weights is never been called anymore.

Thanks.

**Comment by bstriner** [comment]
You should do some reading on how theano and tensorflow build graphs. `get_update` builds a graph that can be used to get updates, it doesn't actually perform the updates. Those updates are then run by the training function.

It is a weird way to start thinking about things. Python code builds the graph and then tells the GPU each time it needs to be run.

The accumulator is created locally by get_updates and initialized to zeros. get_updates is called once. Each time the updates returned from get_updates are run, they update that same local variable.

Cheers

**Comment by zhengsx** [comment]
@bstriner Thank you. The graph building is really confused me. If I want to add a list to the optimizer, the index must be int type not tensor, but if I convert a tensor(which ndim = 0) to int, then the int variable can not update through the iteration. What can I do?

For Example: in `get_update` I define
`
temp_list = [K.zeros(shape=(10, 10)) for i in range(max_iteration)]

self.updates.append(K.update_add(self.iterations, 1))

new_l = K.ones((shape=(10,10)))

self.updates.append(K.update(np.sum(np.asarray(temp_list[int()].eval()))), new_l))
`
The member in list (except temp_list[0]) will not be update because the index (which is int type) can not be updated in the loop.

What should I do?

Thanks.

**Comment by bstriner** [comment]
What are you trying to do? The code makes no sense.

If you're trying to create a new variable and update it as part of training, then it needs to be a theano/tensorflow shared variable and you need to research how to do that. You can't be using numpy in your training function and expect it to run each time it trains. That's why theano and tensorflow have equivalents for most numpy functions.

**Comment by zhengsx** [comment]
I understand. Thank you.

**Comment by bstriner** [comment]
This is a skeleton version of Keras that might be a good explanation:

https://gist.github.com/bstriner/e1e011652b297d13b3ac3f99fd11b2bc

Updates are built into a train function then that function is run. Any variable updates you want to make need to be added to that function and they will run on the GPU.

The train function:
* parameters are inputs and targets
* outputs are loss and other metrics
* performs updates when executed

```python
opt = Adam()
updates = opt.get_updates(params, [], loss)
train = K.function([x, ytrue],[loss, accuracy],updates=updates)
...
for epoch in range(1000):
	loss, accuracy = train([xtrain, ytrain])
